feat: Null coalesce operator

Kamirus opened the pull request:
Closes #5144 

Reduces verbosity of common expressions like
```motoko
switch(opt) {
  case (null) { defaultOrTrap(...) };
  case (?v) { v };
};
```
To just
```motoko
opt ?? defaultOrTrap(...)
```

The operator `??` is commonly used in other mainstream languages like TS/JS and C# so it is familiar for both humans and AI. 

Follow up: @rvanasa @christoph-dfinity we would need to update the grammar for tools like the formatter, lintoko, vscode etc, yes?

Kamirus committed:
feat: Null coalesce operator

Kamirus committed:
fix grammar and add todos

Kamirus committed:
fix astjs

Kamirus committed:
Merge branch 'master' into kamil/null-coalesce-operator

Kamirus committed:
infer + check for ?? and tests

Kamirus committed:
Merge branch 'master' into kamil/null-coalesce-operator

Kamirus committed:
do with ??

Kamirus committed:
review

crusso commented on src/mo_frontend/parser.mly:
```diff
@@ -733,6 +739,8 @@ exp_un(B) :
     { AndE(e1, e2) @? at $sloc }
   | e1=exp_bin(B) OR e2=exp_bin(ob)
     { OrE(e1, e2) @? at $sloc }
+  | e1=exp_bin(B) NULLCOALESCE e2=exp_bin(ob)
```
Since this is a bit like a conditional, we could also consider allowing a block on the rhs (like an else branch).
```
| e1=exp_bin(B) NULLCOALESCE e2=exp_next
```
Maybe a bad idea

Kamirus replied:
Oh! Maybe allowing a block on the left? so that
```motoko
do ? {
} ?? default
```
Would parse properly without a parenthesis? 
Not allowing a block on the RHS would parse records better yes? Without extra brackets

crusso commented:
Looks nice enough to me - lets see what the others think. 

Are there any other (commonly used) symbols for `??` that avoid the parser hacks? `
Perhaps ` e !/ e` (since we already !  in do ? { ... e! })

alexandru-uta commented:
> Are there any other (commonly used) symbols for `??` that avoid the parser hacks? `Perhaps` e !/ e`(since we already use`!` in`do ? { ... e ! ... }`)`.

Since we try to help an AI write more concise code we should stick to `??` because then we can take advantage of the base model knowing JS/TS.

Kamirus commented:
TODO:
Should we allow blocks on the LHS or RHS? 

Examples:
RHS without block:
```motoko
dailyLogs.get(caller) ?? do {
  dailyLogs.add(caller, Map.empty<Nat, DailyRecord>());
  getDailyLogs(caller);
};
```

RHS without block vs with block
```motoko
foo.bar(qux) ?? { x = 0; z = null };
```
vs
```motoko
foo.bar(qux) ?? { { x = 0; z = null } };
```

LHS with block:
```motoko
do ? {
  ...
} ?? default
```
Does it work?

Kamirus committed:
allow block on RHS

Kamirus committed:
fix grammar

Kamirus committed:
fix

Kamirus committed:
Merge branch 'master' into kamil/null-coalesce-operator

Kamirus committed:
fix release

Kamirus committed:
remove macos-15-intel from release

Kamirus commented:
> > I was not aware of the >> disambiguation
> 
> Right sorry, I should've linked it.

Thanks! No worries I would have found it 😄 I meant that I was not aware of that mechanism when I first added the `??` token 😅

Kamirus committed:
Merge branch 'master' into kamil/null-coalesce-operator

Kamirus committed:
disambiguate in the lexer

Kamirus committed:
Merge branch 'master' into kamil/null-coalesce-operator

Kamirus committed:
fix: update error handling for type inconsistencies

Modified the `error_inconsistent_type` function to include spans for better error reporting. Updated the error code from M0240 to M0245 to reflect the changes. Additionally, added new test cases for the null-coalesce operator to validate the expected behavior and error messages.

Kamirus committed:
more tests

Kamirus committed:
up

Kamirus committed:
Merge branch 'master' into kamil/null-coalesce-operator

Kamirus committed:
Merge branch 'master' into kamil/null-coalesce-operator

Kamirus committed:
review: backward-compat unary `??`, error context, changelog

- parser: re-add unary NULLCOALESCE rules in typ_un/exp_un/pat_un so
  `?? T`, `?? e`, `?? p` (whitespace-separated) keep parsing as `?(?X)`
  after the lexer started emitting a single NULLCOALESCE token.
- typing: thread operator context through expect_option_type so M0065
  reports "expected option type before '!'" / "before '??'" instead of
  the bare message.
- error_codes: register M0245 (no best common type for `??` branches);
  refresh M0065 description.
- changelog: document the new operator and its whitespace rules.
- tests: cover unary `??` in type/expr/pattern positions; refresh
  snapshots for M0065 wording.

Made-with: Cursor

Kamirus committed:
fix run test, refresh grammar.txt, tighten typing helpers

- test/run/null-coalesce.mo: parenthesize `(x ?? d) == y` assertions.
  `??` is at exp_nondec precedence so `x ?? 0 == 1` was parsing as
  `x ?? (0 == 1)`, producing M0096 errors that were captured as the
  expected snapshot — meaning the runtime test never executed. Drop the
  stale `tc.ret.ok` / `tc.ok` snapshots so the test now runs through tc,
  ir, low, and wasm-run phases.
- Add a positive case for `a?? b` (lexer only requires whitespace AFTER
  the second `?` to emit NULLCOALESCE).
- doc/md/examples/grammar.txt: regenerate so the published grammar
  includes the unary `'??' <typ_un|exp_un|pat_un>` backward-compat rules.
- typing.ml: inline `expect_option_type`'s call into a single helper and
  drop the unused `error_expected_type` extraction; both shrink the diff
  vs master without changing behavior.
- Changelog: correct the whitespace claim (only required after `??`) and
  call out that `??` binds looser than `==` / arithmetic.

Made-with: Cursor

Kamirus committed:
cleanup lexer disamb
